src/main/java/com/iota/iri/storage/Tangle.java

```diff
@@ -82,13 +83,38 @@ public void shutdown() throws Exception {
      * @see PersistenceProvider#get(Class, Indexable)
      */
     public Persistable load(Class<?> model, Indexable index) throws Exception {
-            Persistable out = null;
-            for(PersistenceProvider provider: this.persistenceProviders) {
-                if((out = provider.get(model, index)) != null) {
-                    break;
+        LinkedList<Persistable> outlist = new LinkedList<>();
+        for (PersistenceProvider provider : this.persistenceProviders) {
+            Persistable result = provider.get(model, index);
+
+            if (result != null && result.exists()) {
+                if (result.merge()) {
+
+                    outlist.add(result);
+                } else {
+                    // If it is a non-mergeable result then there is no need to ask another provider again.
+                    // return immediately
+                    return result;
                 }
             }
-            return out;
+        }
+        Persistable p = outlist.stream().reduce(null, (a, b) -> {
```

Comment:
If we were to cache the persistables, merging them would result in modification of the original. 
I  think its best to change the null into 'model.newInstance()'

Reply:
Is caching seriously considered now? Because there are a lot of arguments to be made to not do index caching (addresses, tags, bundles etc). Is it relevant now?

Reply:
Well...
Currently it is evicted from our priority queue.
However, it is a good practice imo to write robust code that won't break.

We can come to a compromise where we write a comment on the caching issue so that if we ever start working on it we won't forget this change.

Btw, feel free to be explicit aout your arguments against caching

Reply:
Sure! There I go:

Currently, indexes are not being updated after deletion of the underlying data through snapshotting which is described here: https://github.com/iotaledger/iri/issues/1307.

So for any caching to represent data I think this is important to have fixed first. Then on caching: transaction caching is easy because the underlying data won't change. While for 'canMerge'/mergable persistables this can change. There are basically two approaches you can go here when an update arrives on the index(address/tags etc). 

1) Update it on disk and in memory/cache at the same time. 
2) Just evict it from the cache and update on disk.

Both shouldn't cause any trouble with that code.

Another problem would come from the ViewModel .store function and then the tangle no longer knows where to update this. This is generally true but: The .store function is never used, anywhere, not even in tests. Also imo a ViewModel should not have a store function (it is a View after all), that it is not used anywhere points to me this was not really intended either.

Indexes are a function of when a transaction is stored or deleted. Because of this, I think there will be no problem whatsoever with this code in the future. Even with caching.

Reply:
Note that transaction metadata like `solidity` or `snapshotIndex` can change.

I think that if an update arrives on you can just update the cache. There will be a discrepancy between the disk and cache, but this won't matter too much. The cache should be evicted to disk in a way so that if there is a failure, then upon restarting the node should still be able to sync correctly and obtain from its neighbors all the lost data.

I agree with what you said about the `ViewModel`, and in fact I want to get rid of it entirely.

Reply:
So solidity and snapshotIndex are non-mergable persistables. (the only mergable persistables are indexes).

So when it comes to caching, it is a 'problem' on its own and not related to the 'mergablity' of an object?

PS When it comes to caching: updating on disk and just evicting from the cache and then repopulate the cache when it is required -for me- has always worked like a charm. There is some overhead but the simple approach usually makes it very manageable code-wise because there are much fewer edge-cases to handle.